src/ir/subtypes.h

```diff
@@ -73,6 +73,50 @@ struct SubTypes {
     return ret;
   }
 
+  // Computes the depth of children for each type. This is 0 if the type has no
+  // subtypes, 1 if it has subtypes but none of those have subtypes themselves,
+  // and so forth.
+  //
+  // This depth ignores bottom types.
+  std::unordered_map<HeapType, Index> getMaxDepths() {
+    std::unordered_map<HeapType, Index> depths;
+
+    // Begin with depth 0.
+    for (auto type : types) {
+      depths[type] = 0;
+    }
+
+    // Begin with a plan to work on all the types. When we visit an item, we'll
+    // update our super type based on our current depth.
+    std::unordered_set<HeapType> work(types.begin(), types.end());
+
+    while (!work.empty()) {
+      auto iter = work.begin();
+      auto type = *iter;
+      work.erase(iter);
+      if (auto super = type.getSuperType()) {
+        auto depth = depths[type];
+        auto& superDepth = depths[*super];
+        if (depth + 1 >= superDepth) {
```

Comment:
If the `superDepth` doesn't change, then we shouldn't need to keep working.
 ```suggestion
        if (depth + 1 > superDepth) {
```